### PR TITLE
Added parallel inference support to FrameSet

### PR DESCRIPTION
**Pull request by cat101**
Added a threadpool to allow concurrent inference requests. Parallelism, allows us to use LandingEdge pipelining ability (i.e. decode and process images while running inference on GPU). It also maximizes throughput on devices with multiple GPUs.

As part of this work I had to modify the Timer class to be able to support concurrency

### CODE REVIEW

**Comment by srcolinas** [comment]
Are you adding some tests?

**Comment by cat101** [comment]
> Are you adding some tests?

Good catch. I added one to exercise the new code

File: landingai/pipeline/frameset.py [hunk 1]
```diff
@@ -413,15 +417,27 @@ def is_empty(self) -> bool:
         """
         return not self.frames  # True if the list is empty
 
-    def run_predict(self, predictor: Predictor) -> "FrameSet":
+    def run_predict(self, predictor: Predictor, pipelining: bool = False) -> "FrameSet":
```
Comment (pappacena) [inline landingai/pipeline/frameset.py:420]: Can we rename this `pipeelining` parameter to `max_workers: int = 1` instead? 

This way, we control directly if we want parallelization or not, and how many at a time (if `max_workers=1`, we don´t use the threadpool).

File: landingai/pipeline/frameset.py [hunk 2]
```diff
@@ -141,13 +142,16 @@ def from_array(cls, array: np.ndarray, is_bgr: bool = True) -> "Frame":
         im = Image.fromarray(array)
         return cls(image=im)
 
-    def run_predict(self, predictor: Predictor) -> "Frame":
+    def run_predict(self, predictor: Predictor, reuse_session: bool = False) -> "Frame":
```
Comment (pappacena) [inline landingai/pipeline/frameset.py:145]: Should this `reuse_session` be True by default, to keep the previous behavior?

File: landingai/pipeline/frameset.py [hunk 3]
```diff
@@ -141,13 +142,16 @@ def from_array(cls, array: np.ndarray, is_bgr: bool = True) -> "Frame":
...
         """Run a cloud inference model
         Parameters
         ----------
         predictor: the model to be invoked.
+        reuse_session
+            Whether to reuse the session for sending multiple inference requests. By default, the session is reused to improve the performance. If you want to send multiple requests in parallel, set this to False.
```
Comment (pappacena) [inline landingai/pipeline/frameset.py:151]: Can we rephrase this to `Whether to reuse the same HTTP session (...)`  and `If you are sending requests from multiple threads, set this to False`.

File: landingai/pipeline/frameset.py (hunk 1 above)
Comment (cat101) [reply landingai/pipeline/frameset.py:420]: I like the suggestion....I don't like the name...(too technical and void of domain meaning).... still I don't have a better name.. :(

File: landingai/pipeline/frameset.py (hunk 2 above)
Comment (cat101) [reply landingai/pipeline/frameset.py:145]: No, if you are using cloud inference and not calling concurrently, a single session will maintain the connection opened and save SSL negotiation time (i.e. 100s of ms per call)...  A future improvement would be to maintain a pool of connections.

**Review by pappacena — APPROVED** [review]
